### PR TITLE
stage2: only initialize the llvm backend for the target we are building

### LINKED ISSUE #9596: only call LLVM initialization functions for the actual target
Currently, our LLVM bindings look like this:

https://github.com/ziglang/zig/blob/2f1abd919a8465e515875cd08816778df795aabe/src/codegen/llvm/bindings.zig#L683-L780

This is called here:

https://github.com/ziglang/zig/blob/2f1abd919a8465e515875cd08816778df795aabe/src/codegen/llvm.zig#L259-L265

Presumably LLVM breaks these functions up by target because doing the initialization for unused targets is wasted time.

Therefore, this logic should be changed to observe the Target that zig is building for, and only call the appropriate LLVM initialization functions.

An implementation of this issue should come with a quick benchmark test results so we can see if this made any difference.

### PR DESCRIPTION
**Pull request by nektro**
Fixes #9596

### CODE REVIEW

**Review by andrewrk — CHANGES_REQUESTED** [review]
Implementation looks good but please put `initializeLLVMTarget` in codegen/llvm.zig rather than in the bindings file. The reason `LLVMInitializeAllTargetInfos` was in the bindings file was that it was translated from an inline function in an .h file.

File: src/codegen/llvm/bindings.zig [hunk 1]
```diff
@@ -680,103 +685,108 @@ extern fn LLVMInitializeSystemZAsmParser() void;
...
+            LLVMInitializeSystemZAsmParser();
+        },
+        .i386, .x86_64 => {
+            LLVMInitializeX86Target();
+            LLVMInitializeX86TargetInfo();
+            LLVMInitializeX86TargetMC();
+            LLVMInitializeX86AsmPrinter();
+            LLVMInitializeX86AsmParser();
+        },
+        else => {},
```
Comment (andrewrk) [inline src/codegen/llvm/bindings.zig:788]: list all the targets exhaustively please

**Comment by nektro** [comment]
Should I make all the `LLVMInitialize` functions public then?

File: src/codegen/llvm.zig [hunk 2]
```diff
@@ -1957,3 +1949,145 @@ pub const FuncGen = struct {
...
+            llvm.LLVMInitializeX86TargetMC();
+            llvm.LLVMInitializeX86AsmPrinter();
+            llvm.LLVMInitializeX86AsmParser();
+        },
+        .xcore => {
+            llvm.LLVMInitializeXCoreTarget();
+            llvm.LLVMInitializeXCoreTargetInfo();
+            llvm.LLVMInitializeXCoreTargetMC();
+            llvm.LLVMInitializeXCoreAsmPrinter();
+            // llvm.LLVMInitializeXCoreAsmParser();
```
Comment (andrewrk) [inline src/codegen/llvm.zig:2072]:
```suggestion
            // There is no LLVMInitializeXCoreAsmParser function.
```

File: src/codegen/llvm/bindings.zig [hunk 3]
```diff
@@ -1,5 +1,6 @@
 //! We do this instead of @cImport because the self-hosted compiler is easier
 //! to bootstrap if it does not depend on translate-c.
+const std = @import("std");
```
Comment (andrewrk) [inline src/codegen/llvm/bindings.zig:3]: unused import

File: src/codegen/llvm.zig [hunk 4]
```diff
@@ -1957,3 +1949,145 @@ pub const FuncGen = struct {
         return self.llvmModule().getIntrinsicDeclaration(id, null, 0);
     }
 };
+
+pub fn initializeLLVMTarget(arch: std.Target.Cpu.Arch) void {
```
Comment (andrewrk) [inline src/codegen/llvm.zig:1953]:
```suggestion
fn initializeLLVMTarget(arch: std.Target.Cpu.Arch) void {
```

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks - almost there

**Review by andrewrk — APPROVED** [review]

**Comment by marler8997** [comment]
Out of curiosity, what was the performance difference?

**Comment by nektro** [comment]
`773,734 ns` to `90,751 ns` on my `AMD Ryzen 5 4500U (6) @ 2.375GHz`

**Pull request merged**